src/packages/client/src/__tests__/integration/happy/long-running-transaction/test.ts

```diff
@@ -0,0 +1,374 @@
+
+            Invalid \`prisma.user.findMany()\` invocation:
+
+
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.
+          `)
+  })
+
+  /**
+   * Transactions should fail after the changed timeout
+   */
+  test('timeout override', async () => {
+    const result = prisma.$transaction(
+      async (prisma) => {
+        await prisma.user.create({
+          data: {
+            email: 'user_1@website.com',
+          },
+        })
+
+        await new Promise((res) => setTimeout(res, 600))
+
+        return prisma.user.findMany()
+      },
+      {
+        maxWait: 200,
+        timeout: 500,
+      },
+    )
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.findMany()\` invocation:
+
+
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.
+          `)
+  })
+
+  /**
+   * Transactions should fail and rollback if thrown within
+   */
+  test('rollback throw', async () => {
+    const result = prisma.$transaction(async (prisma) => {
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      throw new Error('you better rollback now')
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(
+      `you better rollback now`,
+    )
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * A transaction should fail if it's called inside another transaction
+   */
+  test('nested', async () => {
+    const result = prisma.$transaction(async (tx) => {
+      await tx.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      await prisma.$transaction(async (prisma) => {
+        await prisma.user.create({
+          data: {
+            email: 'user_2@website.com',
+          },
+        })
+      })
+
+      return tx.user.findMany()
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.findMany()\` invocation:
+
+
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.
+          `)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * We don't allow certain methods to be called in a transaction
+   */
+  test('forbidden', async () => {
+    const forbidden = ['$connect', '$disconnect', '$on', '$transaction', '$use']
+
+    const result = prisma.$transaction((prisma) => {
+      // we accumulate all the forbidden methods and expect to be undefined
+      return forbidden.reduce((acc, item) => acc ?? prisma[item], undefined)
+    })
+
+    await expect(result).resolves.toBe(undefined)
+  })
+
+  /**
+   * If one of the query fails, all queries should cancel
```

Comment:
Add a test for query failing with bad input (aka not sent to the engine).

Reply:
Ok, added to my todo list.